test/module/libs/common/result_test.cpp

```diff
@@ -0,0 +1,154 @@
+ * @when match function is invoked
+ * @then match executes successful case
+ */
+TEST(ResultTest, ResultSameTypeValueConstruction) {
+  Result<int, int> result = makeValue(5);
+  result.match([](Value<int> value) { SUCCEED(); },
+               [](Error<int> error) { FAIL() << "Unexpected error case"; });
+}
+
+/**
+ * @given Result constructed with the same type of value and error
+ * @when match function is invoked
+ * @then match executes error case
+ */
+TEST(ResultTest, ResultSameTypeErrorConstruction) {
+  Result<int, int> result = makeError(10);
+  result.match([](Value<int> value) { FAIL() << "Unexpected success case"; },
+               [](Error<int> error) { SUCCEED(); });
+}
+
+/**
+ * @given Two functions which return result
+ * @when bind operator is used to chain these 2 functions and they both return
+ *       valid values
+ * @then result of a bind contains valid value.
+ */
+TEST(ResultTest, ResultBindOperatorSuccesfulCase) {
+  auto get_int = []() -> Result<int, std::string> { return makeValue(10); };
+  auto negate_int = [](int a) -> Result<int, std::string> {
+    return makeValue(-1 * a);
+  };
+
+  auto result = get_int() | negate_int;
+
+  result.match([](Value<int> v) { ASSERT_EQ(-10, v.value); },
+               [](Error<std::string> e) { FAIL() << "Unexpected error case"; });
+}
+
+// function which must not be called for test to pass
+auto never_used_negate_int = [](int a) -> Result<int, std::string> {
+  EXPECT_TRUE(false);
+  return makeValue(-1 * a);
+};
+
+auto error_get_int = []() -> Result<int, std::string> {
+  return makeError("first function");
+};
+
+/**
+ * @given two functions which return result
+ * @when bind operator is used to chain these 2 functions and first one return
+ *       error
+ * @then result of bind contains error and second function is not invoked
+ */
+TEST(ResultTest, ResultBindOperatorErrorFirstFunction) {
+  auto result = error_get_int() | never_used_negate_int;
+
+  result.match(
+      [](Value<int> v) { FAIL(); },
+      [](Error<std::string> e) { ASSERT_EQ("first function", e.error); });
+}
+
+/**
+ * @given two functions which return different results, but value and error
+ *        types can be cast to each other
+ * @when bind operator is used to chain these 2 functions
+ * @then result type is properly deduced
+ */
+TEST(ResultTest, ResultBindOperatorCompatibleTypes) {
+  auto result = error_get_int() | never_used_negate_int;
+
+  static_assert(
+      std::is_same<decltype(result), decltype(never_used_negate_int(1))>::value,
+      "Result type does not match function return type");
+
+  result.match(
+      [](Value<int> v) { FAIL(); },
+      [](Error<std::string> e) { ASSERT_EQ("first function", e.error); });
+}
+
+/**
+ * @given Result with void value
+ * @when match function is invoked
+ * @then void type is correctly handled by compiler
+ */
+TEST(ResultTest, ResultVoidValue) {
+  Result<void, std::string> result = makeError("error message");
```

Comment:
I think it's better prohibit result of void
@nickaleks, @lebdron, @muratovv what do you think?

Reply:
yes, you should not be able to create V=void, but you can create void*

Reply:
Any reasons why is it better?

Reply:
Result<void> can be used to when you need to know if function was successful, and if it wasn't - know the reason of failure

Reply:
@Warchant, well, probably `void*` isn't really suitable for us, we have interfaces in such a case. Nevertheless it's the different type.

@lebdron, I don't really sure there's any use-cases. And that probably may lead some problems in future as long as `void` is a quite special type for templates

@nickaleks, Instead of `Result<void, error>` it's possible use just an error, I see no problem here. Also as [I may guess](https://ideone.com/dxexFo) it's not possible to create an empty value type in this case

Reply:
And what to put in case there is no error? Default value? This is hard to work with